packages/mobile/src/pincode/authentication.ts

```diff
@@ -0,0 +1,99 @@
+import { sha256 } from 'ethereumjs-util'
+import { asyncRandomBytes } from 'react-native-secure-randombytes'
+import { ensureCorrectPin } from 'src/pincode/utils'
+import { store } from 'src/redux/store'
+import { retrieveStoredItem, storeItem } from 'src/storage/keychain'
+import Logger from 'src/utils/Logger'
+import { currentAccountSelector } from 'src/web3/selectors'
+
+const TAG = 'pincode/authentication'
+enum STORAGE_KEYS {
+  PEPPER = 'PEPPER',
+  PASSWORD_HASH = 'PASSWORD_HASH',
+}
+const PEPPER_LENGTH = 64
+export const PIN_LENGTH = 6
+
+let cachedPepper: string
+const cachedPasswordHashes: {
+  [account: string]: string
+} = {}
+
+export function isPinValid(pin: string) {
+  return pin.length === PIN_LENGTH
+}
+
+function pinStorageKey(account: string) {
+  return `${STORAGE_KEYS.PASSWORD_HASH}-${account}`
+}
+
+async function retrieveOrGeneratePepper() {
+  if (cachedPepper) {
+    return cachedPepper
+  }
+  const storedPepper = await retrieveStoredItem(STORAGE_KEYS.PEPPER)
```

Comment:
Should we have error handling at some layer or is it intentional to allow them to propagate up to the component?